File: packages/backend/src/modules/weather/weather.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { ConfigService } from '@nestjs/config';
import { Cron, CronExpression } from '@nestjs/schedule';
import { Prisma } from '@prisma/client';
import { PrismaService } from '@/prisma/prisma.service';
import { WeatherQueryDto, IngestWeatherDto } from './dto/weather.dto';
import { firstValueFrom } from 'rxjs';

@Injectable()
export class WeatherService {
  private readonly logger = new Logger(WeatherService.name);
  private readonly openweatherUrl = 'https://api.openweathermap.org/data/2.5';

  constructor(
    private prisma: PrismaService,
    private http: HttpService,
    private config: ConfigService,
  ) {}

  async getCurrent(query: WeatherQueryDto) {
    const apiKey = this.config.get<string>('WEATHER_API_KEY');

    if (apiKey && (query.lat || query.lng)) {
      try {
        return await this.fetchFromOpenWeather(query);
      } catch (err) {
        this.logger.warn(`OpenWeatherMap fetch failed, trying DB cache: ${err}`);
      }
    }

    const record = await this.prisma.weatherRecord.findFirst({
      where: {
        region: { contains: query.region, mode: 'insensitive' },
        expiresAt: { gt: new Date() },
      },
      orderBy: { recordedAt: 'desc' },
    });

    return record || this.generateSimulatedWeather(query);
  }

  async getForecast(region: string) {
    const apiKey = this.config.get<string>('WEATHER_API_KEY');
    if (apiKey) {
      try {
        const { data } = await firstValueFrom(
          this.http.get(`${this.openweatherUrl}/forecast`, {
            params: { q: region, appid: apiKey, units: 'metric', cnt: 7 },
          }),
        );
        return this.mapForecastResponse(data);
      } catch { this.logger.warn('OpenWeather forecast failed, using simulated'); }
    }

    const record = await this.prisma.weatherRecord.findFirst({
      where: {
        region: { contains: region, mode: 'insensitive' },
        forecast: { not: Prisma.JsonNull },
      },
      orderBy: { recordedAt: 'desc' },
    });
    return record?.forecast || this.generateSimulatedForecast();
  }

  async getAlerts(region: string) {
    const apiKey = this.config.get<string>('WEATHER_API_KEY');
    if (apiKey) {
      try {
        const { data } = await firstValueFrom(
          this.http.get(`${this.openweatherUrl}/weather`, {
            params: { q: region, appid: apiKey, units: 'metric' },
          }),
        );
        return this.checkForAlerts(data);
      } catch { this.logger.warn('OpenWeather alerts check failed'); }
    }

    const record = await this.prisma.weatherRecord.findFirst({
      where: {
        region: { contains: region, mode: 'insensitive' },
        alerts: { not: Prisma.JsonNull },
        expiresAt: { gt: new Date() },
      },
      orderBy: { recordedAt: 'desc' },
    });
    return record?.alerts || [];
  }

  async ingest(dto: IngestWeatherDto) {
    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + 3);
    return this.prisma.weatherRecord.create({ data: { ...dto, expiresAt } });
  }

  private async fetchFromOpenWeather(query: WeatherQueryDto) {
    const params: any = { appid: this.config.get('WEATHER_API_KEY'), units: 'metric' };
    if (query.lat && query.lng) {
      params.lat = query.lat;
      params.lon = query.lng;
    } else {
      params.q = query.region;
    }

    const { data } = await firstValueFrom(
      this.http.get(`${this.openweatherUrl}/weather`, { params }),
    );

    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + 3);

    const record = await this.prisma.weatherRecord.create({
      data: {
        region: data.name || query.region,
        temperature: data.main?.temp ?? 20,
        humidity: data.main?.humidity ?? 50,
        rainfall: data.rain?.['1h'] || data.rain?.['3h'] || 0,
        windSpeed: data.wind?.speed ?? 0,
        sunlightHours: this.estimateSunlight(data),
        condition: this.mapWeatherCondition(data.weather?.[0]?.main),
        forecast: await this.fetchForecastForDb(data.coord?.lat, data.coord?.lon),
        expiresAt,
      },
    });

    return record;
  }

  private async fetchForecastForDb(lat?: number, lon?: number) {
    if (!lat || !lon) return undefined;
    try {
      const { data } = await firstValueFrom(
        this.http.get(`${this.openweatherUrl}/forecast`, {
          params: { lat, lon, appid: this.config.get('WEATHER_API_KEY'), units: 'metric', cnt: 5 },
        }),
      );
      return this.mapForecastResponse(data);
    } catch {
      return undefined;
    }
  }

  private mapWeatherCondition(owmCondition?: string): string {
    const map: Record<string, string> = {
      'Clear': 'CLEAR', 'Clouds': 'CLOUDY', 'Rain': 'RAIN',
      'Drizzle': 'RAIN', 'Thunderstorm': 'STORM', 'Snow': 'RAIN',
      'Mist': 'CLOUDY', 'Fog': 'CLOUDY', 'Haze': 'CLOUDY',
      'Dust': 'DROUGHT', 'Smoke': 'DROUGHT',
    };
    return map[owmCondition || ''] || 'CLEAR';
  }

  private mapForecastResponse(data: any) {
    if (!data?.list) return [];

    const daily = new Map<string, any>();
    for (const item of data.list) {
      const date = item.dt_txt?.split(' ')[0];
      if (!date || daily.has(date)) continue;
      daily.set(date, {
        date,
        temperature: { min: item.main?.temp_min, max: item.main?.temp_max },
        humidity: item.main?.humidity,
        condition: this.mapWeatherCondition(item.weather?.[0]?.main),
        precipitation: item.pop ? item.pop * 100 : 0,
        windSpeed: item.wind?.speed,
      });
    }

    return Array.from(daily.values()).slice(0, 7);
  }

  private checkForAlerts(data: any): any[] {
    const alerts: any[] = [];
    const temp = data.main?.temp;

    if (temp && temp > 40) {
      alerts.push({
        type: 'HEATWAVE',
        severity: 'HIGH',
        message: `Extreme heat (${temp}°C) — protect plants from sunburn`,
        startTime: new Date().toISOString(),
      });
    }
    if (temp && temp < 0) {
      alerts.push({
        type: 'FREEZE',
        severity: 'HIGH',
        message: `Freezing temperatures (${temp}°C) — bring sensitive plants indoors`,
        startTime: new Date().toISOString(),
      });
    }
    if (data.wind?.speed > 15) {
      alerts.push({
        type: 'WIND',
        severity: 'MEDIUM',
        message: `Strong winds (${data.wind.speed} m/s) — secure tall plants`,
        startTime: new Date().toISOString(),
      });
    }

    return alerts;
  }

  private estimateSunlight(data: any): number {
    const clouds = data.clouds?.all ?? 50;
    return Math.max(1, Math.round((100 - clouds) / 10));
  }

  private generateSimulatedWeather(query: WeatherQueryDto) {
    return {
      region: query.region,
      temperature: 20 + Math.random() * 10,
      humidity: 40 + Math.random() * 30,
      rainfall: Math.random() * 10,
      windSpeed: Math.random() * 20,
      sunlightHours: 6 + Math.random() * 6,
      condition: 'CLEAR',
      forecast: this.generateSimulatedForecast(),
    };
  }

  private generateSimulatedForecast() {
    const days = [];
    for (let i = 0; i < 5; i++) {
      const date = new Date();
      date.setDate(date.getDate() + i);
      days.push({
        date: date.toISOString().split('T')[0],
        temperature: { min: 12 + Math.random() * 8, max: 20 + Math.random() * 12 },
        humidity: 40 + Math.random() * 30,
        condition: ['CLEAR', 'CLOUDY', 'RAIN', 'PARTLY_CLOUDY'][Math.floor(Math.random() * 4)],
        precipitation: Math.random() * 40,
      });
    }
    return days;
  }
}

File: services/ai/src/services/weather_service.py
from typing import Dict, Optional, Any
import os
import httpx


class WeatherService:
    def __init__(self) -> None:
        self.api_key = os.getenv("WEATHER_API_KEY", "")
        self.base_url = os.getenv(
            "WEATHER_API_BASE_URL",
            "https://api.openweathermap.org/data/2.5",
        )
        self.client: Optional[httpx.AsyncClient] = None

    async def initialize(self) -> None:
        self.client = httpx.AsyncClient(timeout=10.0)

    async def shutdown(self) -> None:
        if self.client:
            await self.client.aclose()

    async def get_current_weather(
        self, lat: float, lon: float
    ) -> Optional[Dict[str, Any]]:
        if not self.api_key or not self.client:
            return self._mock_weather()
        try:
            response = await self.client.get(
                f"{self.base_url}/weather",
                params={
                    "lat": lat,
                    "lon": lon,
                    "appid": self.api_key,
                    "units": "metric",
                },
            )
            response.raise_for_status()
            data = response.json()
            return {
                "temperature_c": data["main"]["temp"],
                "humidity_pct": data["main"]["humidity"] / 100.0,
                "pressure_hpa": data["main"]["pressure"],
                "wind_speed_ms": data["wind"]["speed"],
                "description": data["weather"][0]["description"],
                "clouds_pct": data["clouds"]["all"] / 100.0,
            }
        except Exception:
            return self._mock_weather()

    async def get_forecast(
        self, lat: float, lon: float, days: int = 3
    ) -> Optional[Dict[str, Any]]:
        if not self.api_key or not self.client:
            return self._mock_forecast(days)
        try:
            response = await self.client.get(
                f"{self.base_url}/forecast",
                params={
                    "lat": lat,
                    "lon": lon,
                    "appid": self.api_key,
                    "units": "metric",
                    "cnt": days * 8,
                },
            )
            response.raise_for_status()
            data = response.json()

            daily_forecasts = []
            for i in range(0, len(data["list"]), 8):
                day = data["list"][i]
                daily_forecasts.append({
                    "date": day["dt_txt"].split()[0],
                    "temperature_high_c": day["main"]["temp_max"],
                    "temperature_low_c": day["main"]["temp_min"],
                    "humidity_pct": day["main"]["humidity"] / 100.0,
                    "rainfall_mm": day.get("rain", {}).get("3h", 0),
                })

            total_rainfall = sum(
                f["rainfall_mm"] for f in daily_forecasts
            )

            return {
                "forecasts": daily_forecasts,
                "total_rainfall_forecast_mm": round(total_rainfall, 1),
                "average_temp_c": round(
                    sum(
                        (f["temperature_high_c"] + f["temperature_low_c"]) / 2
                        for f in daily_forecasts
                    )
                    / max(len(daily_forecasts), 1),
                    1,
                ),
            }
        except Exception:
            return self._mock_forecast(days)

    async def get_environmental_context(
        self, lat: float, lon: float
    ) -> Dict[str, Any]:
        weather = await self.get_current_weather(lat, lon)
        forecast = await self.get_forecast(lat, lon, 1)

        if not weather:
            weather = self._mock_weather()
        if not forecast:
            forecast = self._mock_forecast(1)

        return {
            "current": weather,
            "forecast": forecast,
            "context": {
                "is_hot": weather["temperature_c"] > 30,
                "is_cold": weather["temperature_c"] < 5,
                "is_humid": weather["humidity_pct"] > 0.7,
                "is_dry": weather["humidity_pct"] < 0.3,
                "rain_expected": forecast["total_rainfall_forecast_mm"] > 5,
                "frost_risk": weather["temperature_c"] < 2,
                "heat_stress_risk": weather["temperature_c"] > 35,
            },
        }

    def _mock_weather(self) -> Dict[str, Any]:
        import random
        return {
            "temperature_c": round(random.uniform(15, 32), 1),
            "humidity_pct": round(random.uniform(0.3, 0.8), 2),
            "pressure_hpa": round(random.uniform(1000, 1025), 1),
            "wind_speed_ms": round(random.uniform(0, 8), 1),
            "description": "clear sky",
            "clouds_pct": round(random.uniform(0, 0.8), 2),
        }

    def _mock_forecast(self, days: int) -> Dict[str, Any]:
        import random
        forecasts = []
        total_rain = 0
        for d in range(days):
            rain = round(random.uniform(0, 15), 1)
            total_rain += rain
            forecasts.append({
                "date": f"2026-05-{27 + d:02d}",
                "temperature_high_c": round(random.uniform(22, 32), 1),
                "temperature_low_c": round(random.uniform(15, 22), 1),
                "humidity_pct": round(random.uniform(0.4, 0.8), 2),
                "rainfall_mm": rain,
            })

        return {
            "forecasts": forecasts,
            "total_rainfall_forecast_mm": round(total_rain, 1),
            "average_temp_c": round(
                sum(
                    (f["temperature_high_c"] + f["temperature_low_c"]) / 2
                    for f in forecasts
                )
                / max(len(forecasts), 1),
                1,
            ),
        }
